### PR TITLE
nmcli: Add compatibility for new networkmanager library

### LINKED ISSUE #48055: "This module requires NetworkManager glib API" (NetworkManager-glib package obsoleted)
##### SUMMARY
Fedora NetworkManager-glib (Libraries for adding NetworkManager support to applications (old API)), has been removed from the Fedora repo, [more info](https://apps.fedoraproject.org/packages/fedora-obsolete-packages). Please update the API to match with supported (modern) API's.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
nmcli

##### ANSIBLE VERSION
```
ansible 2.7.0
  config file = /home/kjong/git/ansible/ansible.cfg
  configured module search path = [u'/home/kjong/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.15 (default, Oct 15 2018, 15:24:06) [GCC 8.1.1 20180712 (Red Hat 8.1.1-5)]
```

##### CONFIGURATION
```ANSIBLE_PIPELINING(/home/kjong/git/ansible/ansible.cfg) = True
ANSIBLE_SSH_ARGS(/home/kjong/git/ansible/ansible.cfg) = -C -o ControlMaster=auto -o ControlPersist=60s
DEFAULT_BECOME(/home/kjong/git/ansible/ansible.cfg) = True
DEFAULT_BECOME_ASK_PASS(/home/kjong/git/ansible/ansible.cfg) = True
DEFAULT_BECOME_METHOD(/home/kjong/git/ansible/ansible.cfg) = sudo
DEFAULT_BECOME_USER(/home/kjong/git/ansible/ansible.cfg) = root
DEFAULT_CALLBACK_WHITELIST(/home/kjong/git/ansible/ansible.cfg) = [u'profile_roles', u'skippy']
DEFAULT_FORKS(/home/kjong/git/ansible/ansible.cfg) = 16
DEFAULT_HOST_LIST(/home/kjong/git/ansible/ansible.cfg) = [u'/home/kjong/git/ansible/inventory.ini']
DEFAULT_MANAGED_STR(/home/kjong/git/ansible/ansible.cfg) = # Ansible managed
DEFAULT_REMOTE_PORT(/home/kjong/git/ansible/ansible.cfg) = 4443
DEFAULT_STDOUT_CALLBACK(/home/kjong/git/ansible/ansible.cfg) = skippy
DIFF_ALWAYS(/home/kjong/git/ansible/ansible.cfg) = True
DIFF_CONTEXT(/home/kjong/git/ansible/ansible.cfg) = 3
RETRY_FILES_ENABLED(/home/kjong/git/ansible/ansible.cfg) = False
```

##### OS / ENVIRONMENT
Fedora 29.

##### STEPS TO REPRODUCE
The following doesn't work on Fedora 29 anymore because NetworkManager-glib has been removed.
```  - name: configure network interface
    nmcli:
      conn_name: "{{ network.home_lan.domain }}"
      ifname: "{{ network.home_lan.int }}"
      type: ethernet
      ip4: "{{ network.home_lan.ip | ipv4('address/prefix') | first }}"
      gw4: "{{ network.home_lan.gw | ipv4('address') | first }}"
      dns4: "{{ network.home_lan.dns | ipv4('address') | first }}"
      dns4_search: "{{ network.home_lan.domain }} {{ network.home_vpn.domain }}"
      ip6: "{{ network.home_lan.ip | ipv6('address') | first }}"
      gw6: "{{ network.home_lan.gw | ipv6('address') | first }}"
      dns6: "{{ network.home_lan.dns | ipv6('address') | first }}"
      dns6_search: "{{ network.home_lan.domain }} {{ network.home_vpn.domain }}"
      autoconnect: True
      state: present
    when: (network is defined and network.home_lan is defined) and ansible_hostname == "sovereign"
```

##### EXPECTED RESULTS
Compatibility with the supported API's.

##### ACTUAL RESULTS
```
fatal: [sovereign]: FAILED! => {"changed": false, "msg": "This module requires NetworkManager glib API"}

```

### PR DESCRIPTION
**Pull request by torvitas**
##### SUMMARY
fixes #48055, #64646, #63291, #65770 

##### ISSUE TYPE
- Bugfix Pull Request
- Docs Pull Request

##### COMPONENT NAME
nmcli

##### ADDITIONAL INFORMATION
Use the new nmlib if possible, fallback to the old library.

### CODE REVIEW

**Comment by torvitas** [comment]
I am not sure how to fix that errors but I am pretty sure that these tests did not pass without my changes either.

File: lib/ansible/modules/net_tools/nmcli.py [hunk 1]
```diff
@@ -561,16 +565,19 @@
...
-
-    from gi.repository import NetworkManager, NMClient
-    HAVE_NM_CLIENT = True
+    gi.require_version('NM', '1.0')
+    from gi.repository import NM
 except (ImportError, ValueError):
-    NM_CLIENT_IMP_ERR = traceback.format_exc()
-    HAVE_NM_CLIENT = False
+    try:
+        gi.require_version('NMClient', '1.0')
```
Comment (Akasurde) [inline lib/ansible/modules/net_tools/nmcli.py:576]:
```suggestion
        import gi
        gi.require_version('NMClient', '1.0')
```

**Review by Akasurde — CHANGES_REQUESTED** [review]
Fixes : https://github.com/ansible/ansible/issues/65770

File: lib/ansible/modules/net_tools/nmcli.py (hunk 1 above)
Comment (torvitas) [reply lib/ansible/modules/net_tools/nmcli.py:576]: This is already being imported on line 570.

Comment (Akasurde) [reply lib/ansible/modules/net_tools/nmcli.py:576]: I think if you import this here, it will solve the CI problem.

Comment (torvitas) [reply lib/ansible/modules/net_tools/nmcli.py:576]: I do not think this is a good idea.
1) this were to fix a bug this PR is not supposed to fix (shadow change)
2) if this fixes the test, imho the test is broken not the code

**Comment by agurenko** [comment]
Hm, I'm a little bit confused. This was merged back in December, but as of Ansible 2.9.3 (released 6 days ago) I'm still failing to use nmcli module:

The full traceback is:
Traceback (most recent call last):
File "/tmp/ansible_nmcli_payload_1j6882o3/ansible_nmcli_payload.zip/ansible/modules/net_tools/nmcli.py", line 568, in <module>                                                                                
File "/usr/lib64/python3.6/site-packages/gi/__init__.py", line 130, in require_version
raise ValueError('Namespace %s not available' % namespace)
ValueError: Namespace NMClient not available

and ansible message:

"msg": "Failed to import the required Python library (NetworkManager glib API)

Which is the same as mentioned [here](https://github.com/ansible/ansible/pull/62609), but it says that this one in fact should already fix it. I'm completely lost.

**Comment by torvitas** [comment]
> Hm, I'm a little bit confused. This was merged back in December, but as of Ansible 2.9.3 (released 6 days ago) I'm still failing to use nmcli module:
> 
> The full traceback is:
> Traceback (most recent call last):
> File "/tmp/ansible_nmcli_payload_1j6882o3/ansible_nmcli_payload.zip/ansible/modules/net_tools/nmcli.py", line 568, in
> File "/usr/lib64/python3.6/site-packages/gi/**init**.py", line 130, in require_version
> raise ValueError('Namespace %s not available' % namespace)
> ValueError: Namespace NMClient not available
> 
> and ansible message:
> 
> "msg": "Failed to import the required Python library (NetworkManager glib API)
> 
> Which is the same as mentioned [here](https://github.com/ansible/ansible/pull/62609), but it says that this one in fact should already fix it. I'm completely lost.

Are you sure that the library is actually installed?

**Pull request merged**